### ISSUE TITLE
aws_autoscaling_policy uses policyfriendlyname part of ARN as an ID, leading to unexpected results

### ISSUE THREAD

**Issue by dvishniakov** [terraform-providers/terraform-provider-aws#5155]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version
0.11.7

### Affected Resource(s)

aws_autoscaling_policy

### Terraform Configuration Files

```hcl
# Copy-paste your Terraform configurations here - for large Terraform configs,
# please use a service like Dropbox and share a link to the ZIP file. For
# security, you can also encrypt the files using our GPG public key: https://keybase.io/hashicorp
resource "aws_autoscaling_group" "default" {
# ... put here any working configuration
}
resource "aws_autoscaling_policy" "resource_name" {
  name                   = "scale-out"
  autoscaling_group_name = "${aws_autoscaling_group.default.name}"
  adjustment_type        = "ChangeInCapacity"
  policy_type            = "StepScaling"
  metric_aggregation_type = "Average"
  step_adjustment {
    scaling_adjustment          = 1
    metric_interval_lower_bound = 0
  }
}
```

### Expected Behavior

`policyid` should be used as a resource ID, instead of `policyfriendlyname`. See colored legenda in  the https://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html document, part "Amazon EC2 Auto Scaling". Quote from there:
> arn:aws:autoscaling:region:account-id:scalingPolicy:policyid:autoScalingGroupName/groupfriendlyname:policyname/policyfriendlyname

### Actual Behavior

Seems like Terraform uses `policyfriendlyname` part of the aws_autoscaling_policy's ARN.
When changing Terraform's resource name (`resource_name` in the example above) to something else, terraform plan shows to destroy one resource and to create another resource.
Several times in our experience creation of a new resource happened before destroying of an old resource. Since both new and old resources use friendly name (`scale_out` in the example above) as Terraform's resource ID, the final result is: old policy is deleted and new policy is not created (or not listed in ASG). No policy is visible in AWS Console in ASG.

### Steps to Reproduce

1. `terraform apply`
2. Change the Terraform resource name
3. `terraform apply`

**Comment by bflad** [comment]
Hi @dvishniakov 👋  Sorry for the unexpected behavior.

When you are moving resources (as you mention, changing `resource_name` in the example), you should use the [`terraform state mv` command](https://www.terraform.io/docs/commands/state/mv.html) to move the resource in the Terraform state instead of allowing the actual resource to delete and recreate. We cannot perform any proper operation ordering in this race condition scenario as Terraform sees this as two separate resources with separate operations.

Here's an example:

```
$ terraform state mv aws_autoscaling_policy.resource_name_old aws_autoscaling_policy.resource_name_new
```

Hope this helps. 👍 

If you are interested in any enhancements to the Terraform behavior or documentation regarding this (its generic across all resources), I would suggest creating an issue upstream in https://github.com/hashicorp/terraform/issues -- Unfortunately, we cannot provide any changes for this behavior within the AWS provider at this time as there are no hooks available in the Terraform provider SDK to automatically migrate resources.

**Comment by dvishniakov** [comment]
@bflad, just to make sure: is it Terraform's code (not aws provider) which decides to use friendly name instead of policyId?

**Comment by bflad** [comment]
The `aws_autoscaling_policy` resource does in fact choose to perform API operations based on the friendly name instead of the "policy ID" contained within the ARN, but this is mostly due to the design of Autoscaling API (see also the [Autoscaling API Reference](https://docs.aws.amazon.com/autoscaling/ec2/APIReference/)) and should be irrelevant in this scenario compared to the operation it appears you are trying to perform within Terraform.

The `resource_name` is a unique identifier within the Terraform state. When moving a resource configuration within Terraform (changing `resource_name`), you want to ensure that the Terraform state is kept in sync with the configuration change. When Terraform core sees a resource (based on `resource_name`) being removed from a configuration, it will call on the provider to call the resource's delete function. The opposite applies to when Terraform core sees a new resource (based on `resource_name`) being added to a configuration.

In the scenario described above, Terraform core (the parts that handle the dependency graph and ordering of operations) currently has no way of knowing or being told from the provider that you are performing two operations on the same Autoscaling Policy or to automatically inspect the Terraform resource deletion and creation to see if they are the same to potentially handle the change automatically within the Terraform state.

So in this scenario (by not calling `terraform state mv`), `terraform plan` output should be showing something like (the actual plan output here is not indicative of the true order of operations within Terraform):

```
An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create
  - destroy

Terraform will perform the following actions:

  - aws_autoscaling_policy.resource_name_old

  + aws_autoscaling_policy.resource_name_new
... more information here about "new" aws_autoscaling_policy ...
```

There are a few things that make this scenario less than ideal, especially for this particular API:
* The AutoScaling API uses the policy name for creation and updates ([`PutScalingPolicy`](https://docs.aws.amazon.com/autoscaling/ec2/APIReference/API_PutScalingPolicy.html))
* The Autoscaling API uses the policy name for deletion ([`DeletePolicy`](https://docs.aws.amazon.com/autoscaling/ec2/APIReference/API_DeletePolicy.html))
* The EC2 service has an [eventual consistency model](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/query-api-troubleshooting.html#eventual-consistency), so changes may propagate strangely with conflicting  concurrent operations (a PUT and DELETE operation on the same policy name)
* It's probably also worth mentioning that Terraform, by default, runs with up to 10 parallel operations (controllable via the [`terraform apply --parallelism` flag](https://www.terraform.io/docs/commands/apply.html#parallelism-n)) and these two resources are completely separate (in terms of how Terraform's difference logic sees them) so the delete function call and create function call can happen in any order or concurrently

These combined are likely why you are seeing the behavior you describe.

So in short, while the ARN does contain a particular ID value, its lack of usage with API itself prevents Terraform from gaining any benefits of working with it. It also would not necessarily fix the issue described, which should be handled via `terraform state mv`.
